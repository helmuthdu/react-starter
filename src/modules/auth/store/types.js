// Getters (GET*)
// [CONTEXT]_[ACTION]_[PROPERTY]

// Reducer (SET*, ADD*, REMOVE*)
// [CONTEXT]_[ACTION]_[PROPERTY]

// Actions (FETCH*, CREATE*, UPDATE*, DELETE*)
// [CONTEXT]_[ACTION]_[PROPERTY]_[STATUS]
export const AUTH_SET_USER = 'AUTH_SET_USER';
