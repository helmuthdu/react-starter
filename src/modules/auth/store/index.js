export const authStore = require('./modules/auth');
