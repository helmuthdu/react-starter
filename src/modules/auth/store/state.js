// @flow
export interface State {
  username: string;
  email: string;
  isLogged: boolean;
  token: string;
}

export const initialState: State = {
  username: '',
  email: '',
  isLogged: false,
  token: ''
};
