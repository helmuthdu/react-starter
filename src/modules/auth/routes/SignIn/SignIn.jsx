// @flow
import { push } from 'connected-react-router';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createObservableFromInput } from '../../../../helpers/observable';
import { doLogin, doLogout, getUserInfo } from '../../store/modules/auth';

type Props = {
  name: string,
  linkTo: () => void,
  doLogin: () => void,
  getUserInfo: () => void,
  doLogout: () => void
};

class SignIn extends Component<Props> {
  inputField = React.createRef();

  async componentDidMount() {
    await this.props.getUserInfo();
    createObservableFromInput(this.inputField, {}).subscribe((value: any) => {
      console.log(value);
    });
  }

  render() {
    return (
      <form onSubmit={e => e.preventDefault()}>
        <input ref={this.inputField} type="text" placeholder="Username" required />
        <input type="password" placeholder="Password" required />
        <button type="submit">Login</button>
        <p>current user: {this.props.name}</p>
      </form>
    );
  }
}

const mapStateToProps = state => ({ ...state.auth });

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      linkTo: () => push(`/`),
      doLogin,
      getUserInfo,
      doLogout
    },
    dispatch
  );

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SignIn);
