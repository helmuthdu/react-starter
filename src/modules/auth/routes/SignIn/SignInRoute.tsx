import { push } from 'connected-react-router';
import React, { Component, HTMLAttributes, Ref } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, compose, Dispatch } from 'redux';
import { createObservableFromInput } from '../../../../helpers/observable';
import { AppState } from '../../../../index';
import { AuthenticatePayload } from '../../api/auth.api';
import { doLogin, doLogout, getUserData } from '../../store/modules/auth';

type State = {};

type StateProps = {
  name: string;
};

type DispatchProps = {
  linkTo: () => void;
  doLogin: (payload: AuthenticatePayload) => void;
  getUserInfo: () => void;
  doLogout: () => void;
};

type OwnProps = HTMLAttributes<HTMLFormElement>;

export type Props = StateProps & DispatchProps & OwnProps;

export class SignInRoute extends Component<Props, State> {
  inputField: Ref<HTMLInputElement> = React.createRef();

  async componentDidMount() {
    await this.props.getUserInfo();
    createObservableFromInput(this.inputField, {}).subscribe((value: any) => {
      console.log(value);
    });
  }

  render() {
    return (
      <form onSubmit={e => e.preventDefault()}>
        <input ref={this.inputField} type="text" placeholder="Username" required />
        <input type="password" placeholder="Password" required />
        <button type="submit">Login</button>
        <p>current user: {this.props.name}</p>
      </form>
    );
  }
}

const mapStateToProps = (state: AppState): StateProps => ({ name: state.auth.name });

const mapDispatchToProps = (dispatch: Dispatch): DispatchProps =>
  bindActionCreators(
    {
      linkTo: () => push(`/`),
      doLogin,
      getUserInfo: getUserData,
      doLogout
    },
    dispatch
  );

const enhance = compose<React.ComponentType<OwnProps>>(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )
);

export default enhance(SignInRoute);
