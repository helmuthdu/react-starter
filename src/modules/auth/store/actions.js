// @flow
import { authApi, AuthenticatePayload } from '../api/auth.api';
import { AUTH_SET_USER } from './types';

export const getUserInfo = () => async dispatch =>
  dispatch({
    type: AUTH_SET_USER,
    payload: {
      ...(await authApi.get())
    }
  });

export const doLogin = (payload: AuthenticatePayload) => async dispatch => {
  dispatch({
    type: AUTH_SET_USER,
    payload: {
      ...(await authApi.post(payload)).data,
      isLogged: true
    }
  });
};

export const doLogout = () => dispatch =>
  dispatch({
    type: AUTH_SET_USER,
    payload: { username: '', email: '', isLogged: false, token: '' }
  });
