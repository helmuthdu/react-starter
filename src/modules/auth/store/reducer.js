// @flow
import { initialState, State } from './state';
import { AUTH_SET_USER } from './types';

export interface Actions {
  +type: AUTH_SET_USER;
  payload: State;
}

// Reducer
export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case AUTH_SET_USER:
      return { ...state, ...action.payload };
    default:
      return state;
  }
};
