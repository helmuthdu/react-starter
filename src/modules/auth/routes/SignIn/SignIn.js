// @flow
import { push } from 'connected-react-router';
import React, { Component } from 'react';
import { frontloadConnect } from 'react-frontload';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { doLogin, doLogout } from '../../store';

class SignIn extends Component {
  render() {
    return (
      <form onSubmit={e => e.preventDefault()}>
        <input type="text" placeholder="Username" required />
        <input type="password" placeholder="Password" required />
        <button type="submit">Login</button>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  isLogged: state.auth.isLogged
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      changePage: () => push(`/`),
      doLogin,
      doLogout
    },
    dispatch
  );

// Request initial data for the component
const loadRequest = async props =>
  await new Promise((resolve, reject) => {
    setTimeout(() => {
      console.log('LOAD_COMPLETED');
      resolve();
    }, 1000);
  });

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(frontloadConnect(loadRequest)(SignIn));
