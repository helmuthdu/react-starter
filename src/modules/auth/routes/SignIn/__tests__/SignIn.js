import { shallow } from 'enzyme';
import React from 'react';
import configureMockStore from 'redux-mock-store';
import { initialState } from '../../../store';
import SignIn from '../SignIn';

describe('Route -> SignIn component', () => {
  const props = {
    isLogged: false,
    dispatch: jest.fn(),
    changePage: jest.fn(),
    store: configureMockStore()({ auth: initialState })
  };

  it('should match snapshot', () => {
    const wrapper = shallow(<SignIn {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
