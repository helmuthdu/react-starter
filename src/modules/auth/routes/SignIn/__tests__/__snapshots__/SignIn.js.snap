// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route -> SignIn component should match snapshot 1`] = `
<Component
  changePage={[Function]}
  dispatch={[MockFunction]}
  doLogin={[Function]}
  doLogout={[Function]}
  email=""
  getUserInfo={[Function]}
  isLogged={false}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  token=""
  username=""
/>
`;
